refactor(semantic): align handling of declaring symbol for function with TypeScript

Dunqing opened the pull request:
The `Symbol` is based on the `TypeScript` implementation, but `TypeScript` doesn't support non-strict mode. So we tweaked the implementation to support non-strict mode before, but that held us back from adding symbol id for `declare function`, because `declare function oxc(): number; declare function oxc(): void` is valid, but we have no idea except introducing more symbol flags by to check whether it is a declare function. But introducing more symbol flags will lead to greater discrepancies between our implementation and `TypeScript`. https://github.com/oxc-project/oxc/pull/10074 has re-implemented the function redeclaration check in `checker` so that this PR can make a change to align with `TypeScript` implementation without leading to unexpected errors.

A lot of error messages are removed in snapshots, because they are all duplicated, which was introduced by https://github.com/oxc-project/oxc/pull/10074

Dunqing commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10086?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#10078** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10078?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#10086** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10086?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10086?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#10074** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10074?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#10077** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10077?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#10062** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10062?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#10060** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10060?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#10059** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10059?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#10058** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10058?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=10086).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Boshen commented:
### Merge activity

* **Mar 28, 12:11 AM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

Dunqing committed:
refactor(semantic): align handling of declaring symbol for function with TypeScript (#10086)

The `Symbol` is based on the `TypeScript` implementation, but `TypeScript` doesn't support non-strict mode. So we tweaked the implementation to support non-strict mode before, but that held us back from adding symbol id for `declare function`, because `declare function oxc(): number; declare function oxc(): void` is valid, but we have no idea except introducing more symbol flags by to check whether it is a declare function. But introducing more symbol flags will lead to greater discrepancies between our implementation and `TypeScript`. https://github.com/oxc-project/oxc/pull/10074 has re-implemented the function redeclaration check in `checker` so that this PR can make a change to align with `TypeScript` implementation without leading to unexpected errors.

A lot of error messages are removed in snapshots, because they are all duplicated, which was introduced by https://github.com/oxc-project/oxc/pull/10074
